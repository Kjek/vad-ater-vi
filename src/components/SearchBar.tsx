import type { Dispatch, SetStateAction } from 'react';
import { useRef } from 'react';
import { useEffect, useState } from 'react';
import { useTheme } from '~/hooks/useTheme';

interface SearchBarProps {
  setSearchQuery: Dispatch<SetStateAction<string>>;
}

const SearchBar = (props: SearchBarProps) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const [inputText, setInputText] = useState<string>();
  const { theme } = useTheme();
  const { setSearchQuery } = props;

  const onClear = () => {
    if (inputRef.current) {
      inputRef.current.value = '';
      setInputText(undefined);
    }
  };

  useEffect(() => {
    setSearchQuery(inputText ?? '');
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [inputText]);

  return (
    <>
      <div className='flex justify-center'>
        <div className='relative'>
          <input
            title='Sök på restauranger som ska visas i listan'
            type='text'
            id='search-restaurant'
            ref={inputRef}
            className='self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 transition duration-500 focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600 md:w-auto'
            placeholder='Sök restauranger...'
            onChange={(event) => setInputText(event.currentTarget.value)}
          />
          <button
            type='button'
            className='absolute right-0 bottom-[5px] rounded-lg px-2 py-2 text-sm font-medium text-white sm:bottom-[55px] lg:bottom-[5px]'
            onClick={() => onClear()}
          >
            {inputText && (
              <svg
                xmlns='http://www.w3.org/2000/svg'
                viewBox='0 0 24 24'
                fill={theme === 'dark' ? '#767f89' : 'dark'}
                className='h-4 w-4'
              >
                <path
                  fillRule='evenodd'
                  d='M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z'
                  clipRule='evenodd'
                />
              </svg>
            )}
          </button>
        </div>
      </div>
    </>
  );
};

export default SearchBar;
