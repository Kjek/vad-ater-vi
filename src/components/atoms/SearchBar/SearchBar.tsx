import { useTheme } from '@hook/useTheme';
import type { Dispatch, SetStateAction } from 'react';

interface SearchBarProps {
  searchQuery: string;
  setSearchQuery: Dispatch<SetStateAction<string>>;
}

const SearchBar = ({ searchQuery, setSearchQuery }: SearchBarProps) => {
  const { theme } = useTheme();

  const onClear = () => {
    setSearchQuery('');
  };

  return (
    <>
      <div className='hidden justify-center md:flex'>
        <div className='relative w-full sm:w-auto'>
          <input
            title='Sök på restauranger som ska visas i listan'
            type='text'
            id='search-restaurant'
            className='w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none transition duration-500 focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600 md:w-auto'
            placeholder='Sök restauranger...'
            onChange={(event) => setSearchQuery(event.currentTarget.value)}
          />
          <button
            aria-label='rensa sökfältet'
            type='button'
            className='absolute bottom-[5px] right-0 rounded-lg px-2 py-2 text-sm font-medium text-white sm:bottom-[55px] lg:bottom-[5px]'
            onClick={() => onClear()}
          >
            {searchQuery && (
              <svg
                xmlns='http://www.w3.org/2000/svg'
                viewBox='0 0 24 24'
                fill={theme === 'dark' ? '#767f89' : 'dark'}
                className='h-4 w-4'
              >
                <path
                  fillRule='evenodd'
                  d='M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z'
                  clipRule='evenodd'
                />
              </svg>
            )}
          </button>
        </div>
      </div>
    </>
  );
};

export default SearchBar;
