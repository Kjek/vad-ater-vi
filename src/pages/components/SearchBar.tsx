import type { Dispatch, SetStateAction } from 'react';

interface SearchBarProps {
  setSearchQuery: Dispatch<SetStateAction<string>>;
}

const SearchBar = (props: SearchBarProps) => {
  const { setSearchQuery } = props;
  return (
    <>
      <input
        type='text'
        id='search-restaurant'
        className='w-1/2 rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-900 transition duration-500 focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-white dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600 md:w-1/3 lg:w-1/6'
        placeholder='SÃ¶k restauranger...'
        required
        onChange={(event) => setSearchQuery(event.currentTarget.value)}
      />
    </>
  );
};

export default SearchBar;
