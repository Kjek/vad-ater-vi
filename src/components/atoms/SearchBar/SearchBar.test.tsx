import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import SearchBar from './SearchBar';

describe(SearchBar, () => {
  let searchQuery = '';
  const setSearchQuery = vi.fn().mockImplementation((newValue: string) => {
    searchQuery = newValue;
  });

  beforeEach(() => {
    vi.clearAllMocks();
    searchQuery = '';
  });

  it('renders correctly', () => {
    const { getByDisplayValue } = render(
      <SearchBar searchQuery={searchQuery} setSearchQuery={setSearchQuery} />
    );
    const searchBar = getByDisplayValue('') as HTMLInputElement;
    expect(searchBar).toBeInTheDocument();
    expect(searchBar).toBeVisible();
  });

  it('displays value correctly', () => {
    const { getByDisplayValue } = render(
      <SearchBar searchQuery={searchQuery} setSearchQuery={setSearchQuery} />
    );

    const searchBar = getByDisplayValue('') as HTMLInputElement;
    expect(searchBar).toBeVisible();
    expect(searchBar).toHaveValue('');
    expect(searchBar).toHaveValue(searchQuery);
    fireEvent.change(searchBar, { target: { value: 'test' } });
    expect(searchBar).toHaveValue('test');
    expect(searchBar).toHaveValue(searchQuery);
  });
});
