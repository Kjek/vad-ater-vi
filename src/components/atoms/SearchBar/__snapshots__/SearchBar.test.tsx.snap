// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="hidden justify-center md:flex"
      >
        <div
          class="relative w-full sm:w-auto"
        >
          <input
            class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none transition duration-500 focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600 md:w-auto"
            id="search-restaurant"
            placeholder="Sök restauranger..."
            title="Sök på restauranger som ska visas i listan"
            type="text"
          />
          <button
            aria-label="rensa sökfältet"
            class="absolute bottom-[5px] right-0 rounded-lg px-2 py-2 text-sm font-medium text-white sm:bottom-[55px] lg:bottom-[5px]"
            type="button"
          >
            <svg
              class="h-4 w-4"
              fill="dark"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="hidden justify-center md:flex"
    >
      <div
        class="relative w-full sm:w-auto"
      >
        <input
          class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none transition duration-500 focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600 md:w-auto"
          id="search-restaurant"
          placeholder="Sök restauranger..."
          title="Sök på restauranger som ska visas i listan"
          type="text"
        />
        <button
          aria-label="rensa sökfältet"
          class="absolute bottom-[5px] right-0 rounded-lg px-2 py-2 text-sm font-medium text-white sm:bottom-[55px] lg:bottom-[5px]"
          type="button"
        >
          <svg
            class="h-4 w-4"
            fill="dark"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
